modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -3421,24 +3418,81 @@ else if (cntr == maxCntr.cnt)
+
+        if (resetOwners)
+            resetOwnersByCounter(top, maxCntrs, haveHistory);
+
+        return list;
+    }
+
+    /**
+     * Determine new owners for partitions.
+     * If anyone of OWNING partitions have a counter less than maximum this partition changes state to MOVING forcibly.
+     *
+     * @param top Topology.
+     * @param maxCntrs Max counter partiton map.
+     * @param haveHistory Set of partitions witch have historical supplier.
+     */
+    private void resetOwnersByCounter(GridDhtPartitionTopology top,
+        Map<Integer, CounterWithNodes> maxCntrs, Set<Integer> haveHistory) {
+        Map<Integer, Set<UUID>> ownersByUpdCounters = U.newHashMap(maxCntrs.size());
+        Map<Integer, Long> partSizes = U.newHashMap(maxCntrs.size());
+
+        for (Map.Entry<Integer, CounterWithNodes> e : maxCntrs.entrySet()) {
+            ownersByUpdCounters.put(e.getKey(), e.getValue().nodes);
+
+            partSizes.put(e.getKey(), e.getValue().size);
+        }
+
+        top.globalPartSizes(partSizes);
+
+        Map<UUID, Set<Integer>> partitionsToRebalance = top.resetOwners(
+            ownersByUpdCounters, haveHistory, this);
+
+        for (Map.Entry<UUID, Set<Integer>> e : partitionsToRebalance.entrySet()) {
+            UUID nodeId = e.getKey();
+            Set<Integer> parts = e.getValue();
+
+            for (int part : parts)
+                partsToReload.put(nodeId, top.groupId(), part);
+        }
+    }
+
+    /**
+     * Find and assign suppliers for history rebalance.
+     *
+     * @param top Topology.
+     * @param maxCntrs Max counter partiton map.
+     * @param nonMaxCntrs Min counter partiton map.
+     * @param haveHistory Set of partitions witch have historical supplier.
+     * @return List of partitions which does not have historical supplier.
+     */
+    private List<SupplyPartitionInfo> assignHistoricalSuppliers(
+        GridDhtPartitionTopology top,
+        Map<Integer, CounterWithNodes> maxCntrs,
+        Map<Integer, TreeSet<Long>> nonMaxCntrs,
+        Set<Integer> haveHistory
+    ) {
         Map<Integer, Map<Integer, Long>> partHistReserved0 = partHistReserved;
 
         Map<Integer, Long> localReserved = partHistReserved0 != null ? partHistReserved0.get(top.groupId()) : null;
 
-        Set<Integer> haveHistory = new HashSet<>();
-
         List<SupplyPartitionInfo> list = new ArrayList<>();
 
-        for (Map.Entry<Integer, T2<UUID, Long>> e : minCntrs.entrySet()) {
+        for (Map.Entry<Integer, TreeSet<Long>> e : nonMaxCntrs.entrySet()) {
             int p = e.getKey();
-            long minCntr = e.getValue().get2();
 
             CounterWithNodes maxCntrObj = maxCntrs.get(p);
 
             long maxCntr = maxCntrObj != null ? maxCntrObj.cnt : 0;
 
-            // If minimal counter is zero, do clean preloading.
-            if (minCntr == 0 || minCntr == maxCntr)
+            NavigableSet<Long> minCntrs = e.getValue().headSet(maxCntr, false);
```

Comment:
And exactly this map can be named "nonMaxCntrs"

Reply:
Done.